subprojects/base-services/src/main/java/org/gradle/internal/resources/DefaultResourceLockCoordinationService.java

```diff
@@ -82,29 +83,48 @@ public ResourceLockState getCurrent() {
         }
     }
 
-    @Override
-    public void notifyStateChange() {
+    private void maybeNotifyStateChange(DefaultResourceLockState resourceLockState) {
+        if (!resourceLockState.getUnlockedResources().isEmpty()) {
+            notifyStateChange();
+        }
+    }
+
+    private void notifyStateChange() {
         synchronized (lock) {
             lock.notifyAll();
         }
     }
 
-    private void releaseLocks(DefaultResourceLockState stateLock) {
-        for (ResourceLock resourceLock : stateLock.getResourceLocks()) {
+    private void releaseLocks(DefaultResourceLockState resourceLockState) {
+        for (ResourceLock resourceLock : Lists.newArrayList(resourceLockState.getLockedResources())) {
             resourceLock.unlock();
         }
     }
 
     private static class DefaultResourceLockState implements ResourceLockState {
-        private final Set<ResourceLock> resourceLocks = Sets.newHashSet();
+        private final Set<ResourceLock> lockedResources = Sets.newHashSet();
```

Comment:
❌  Afaik this is always called from a single thread, so we can lazily allocate these sets.